vertical-pod-autoscaler/pkg/recommender/input/metrics/metrics_source.go

```diff
@@ -0,0 +1,146 @@
+}
+
+// NewPodMetricsesSource Returns a Source-wrapper around PodMetricsesGetter.
+func NewPodMetricsesSource(source resourceclient.PodMetricsesGetter) MetricsSource {
+	return podMetricsSource{metricsGetter: source}
+}
+
+func (s podMetricsSource) List(ctx context.Context, namespace string, _ *model.ClusterState, opts v1.ListOptions) (*v1beta1.PodMetricsList, error) {
+	podMetricsInterface := s.metricsGetter.PodMetricses(namespace)
+	return podMetricsInterface.List(ctx, opts)
+}
+
+type externalMetricsClient struct {
+	externalClient external_metrics.ExternalMetricsClient
+	options        ExternalClientOptions
+}
+
+// ExternalClientOptions specifies parameters for using an External Metrics Client.
+type ExternalClientOptions struct {
+	CpuMetric, MemoryMetric                          string
+	PodNamespaceLabel, PodNameLabel                  string
+	CtrNamespaceLabel, CtrPodNameLabel, CtrNameLabel string
+}
+
+// NewExternalClient returns a Source for an External Metrics Client.
+func NewExternalClient(c *rest.Config, options ExternalClientOptions) MetricsSource {
+	extClient, err := external_metrics.NewForConfig(c)
+	if err != nil {
+		klog.Fatalf("Failed initializing external metrics client: %v", err)
+	}
+	return externalMetricsClient{
+		externalClient: extClient,
+		options:        options,
+	}
+}
+
+func (s externalMetricsClient) containerId(value externalmetricsv1beta1.ExternalMetricValue) *model.ContainerID {
+	podNS, hasPodNS := value.MetricLabels[s.options.PodNamespaceLabel]
+	podName, hasPodName := value.MetricLabels[s.options.PodNameLabel]
+	ctrName, hasCtrName := value.MetricLabels[s.options.CtrNameLabel]
+	if hasPodNS && hasPodName && hasCtrName {
+		return &model.ContainerID{
+			PodID:         model.PodID{Namespace: podNS, PodName: podName},
+			ContainerName: ctrName,
+		}
+	}
+	return nil
+
+}
+
+type podContainerResourceMap map[model.PodID]map[string]k8sapiv1.ResourceList
+
+func (s externalMetricsClient) addMetrics(list *externalmetricsv1beta1.ExternalMetricValueList, name k8sapiv1.ResourceName, resourceMap *podContainerResourceMap) {
+	for _, val := range list.Items {
+		if id := s.containerId(val); id != nil {
+			(*resourceMap)[id.PodID][id.ContainerName][name] = val.Value
+		}
+	}
+}
+
+func (s externalMetricsClient) List(ctx context.Context, namespace string, state *model.ClusterState, opts v1.ListOptions) (*v1beta1.PodMetricsList, error) {
+	result := v1beta1.PodMetricsList{}
+	// Get all VPAs in the namespace
+	// - We already do this in the cluster state feeder!  It's in its clusterState member.
+	//   We just have to feed it into here somehow.
+	// - use the 'PodSelector' there as the input to the external api.
+	// Send out the queries.
+	nsClient := s.externalClient.NamespacedMetrics(namespace)
+
+	for _, vpa := range state.Vpas {
+		if vpa.PodCount == 0 {
+			continue
+		}
+		workloadValues := make(podContainerResourceMap)
+		cpuMetrics, err := nsClient.List(s.options.CpuMetric, vpa.PodSelector)
```

Comment:
There is no place where you validate that "s.options.CpuMetric" is not an empty value, it is a program parameter. Maybe it is acceptable that the value is empty because we don't want to deal with CPU (or Mem or Net), but in that case we should skip the processing for that particular resource.

Reply:
IMO all along the code we should work with a resourceList (eg: ["cpu"] or ["mem"] or ["cpu","mem","net"] or ....)

In the end you are storing results in a `k8sapiv1.ResourceList` so we really should handle a list of resource, or this code is going to break as soon as a new resource is added in the pipe.